src/main/java/org/folio/rest/support/ShelvingOrderUpdate.java

```diff
@@ -0,0 +1,197 @@
+      .append("LIMIT $1").toString();
+
+
+  private static final String SQL_UPDATE_ITEMS = new StringBuilder()
+      .append("UPDATE ITEM ")
+      .append("SET jsonb = $1 ")
+      .append("WHERE id = $2").toString();
+
+  // Defines version threshold where this property was introduces from
+  private static final String FROM_VERSION_SHELVING_ORDER = "20.1.0";
+
+  private static final String OPERATION_STATUS_VERSION_MISMATCH = "Module version isn't expected";
+  private static final String OPERATION_STATUS_ITEMS_UPDATES_SUCCESSFUL = "Items updates completed successfully";
+  private static final String OPERATION_STATUS_ITEMS_UPDATES_FAILED = "Items updates failed";
+
+  private Versioned versionChecker;
+
+  private int updateBunchSize;
+  private Map<UUID, String> itemsShelvingOrderMap;
+  private Map<UUID, Item> itemsMap;
+
+  public ShelvingOrderUpdate(int updateBunchSize) {
+    this.versionChecker = new Versioned() { };
+    this.versionChecker.setFromModuleVersion(FROM_VERSION_SHELVING_ORDER);
+    this.updateBunchSize = updateBunchSize;
+
+    itemsShelvingOrderMap = new LinkedHashMap<>();
+    itemsMap = new LinkedHashMap<>();
+  }
+
+  public Future<String> updateItems(TenantAttributes attributes, Map<String, String> headers, Context vertxContext) {
+    Promise<String> promise = Promise.promise();
+
+    String fromModuleVersion = attributes.getModuleFrom().trim();
+    if (StringUtils.isBlank(fromModuleVersion) || versionChecker.isNewForThisInstall(fromModuleVersion)) {
+      // Update items only for versions before the expected
+      promise.fail(OPERATION_STATUS_VERSION_MISMATCH);
+    } else {
+      PostgresClient postgresClient =  PgUtil.postgresClient(vertxContext, headers);
+
+      hasItemsToUpdate(vertxContext.owner(), postgresClient)
+          .onSuccess(itemsExists -> {
+            log.warn("Items update started");
+          }).onFailure(h -> {
+            log.info("No items for update");
+      });
+    }
+
+    return promise.future();
+  }
+
+  private Future<Boolean> hasItemsToUpdate(Vertx vertx, PostgresClient postgresClient) {
+    Promise<Boolean> promise = Promise.promise();
+
+    postgresClient.execute(SQL_EXISTS_ITEMS_TO_UPDATE, ar -> {
+      if (ar.succeeded()) {
+        RowSet<Row> rows = ar.result();
+        boolean itemsExists = rows.iterator().next().getBoolean(0);
+        promise.complete(itemsExists);
+
+        fetchAndUpdatesItems(vertx, postgresClient)
+            .onSuccess(h -> hasItemsToUpdate(vertx, postgresClient))
+            .onFailure(h -> log.warn("Current items updates failed: "  + h.getMessage()));
+
+      } else {
+        promise.fail("Failure: " + ar.cause().getMessage());
+      }
+    });
+
+    return promise.future();
+  }
+
+  private Future<Integer> fetchAndUpdatesItems(Vertx vertx, PostgresClient postgresClient) {
+    Promise<Integer> promise = Promise.promise();
+
+    vertx.setTimer(500L, timer -> {
+      itemsShelvingOrderMap.clear();
```

Comment:
This might be unsafe. If the code executed concurrently then the data will be mixed up. Can we make the map local vs holding the state in class?

Reply:
Here I think this extra data structure isn't required at all, it will be removed. All the data required for update will be collected internally in handler as simple list of object with further inline transformation and passing to update routine.